std/mutex.zig

```diff
@@ -21,7 +36,28 @@ pub const Mutex = struct {
     }
 
     pub fn acquire(self: *Mutex) Held {
-        while (@atomicRmw(u8, &self.lock, builtin.AtomicRmwOp.Xchg, 1, AtomicOrder.SeqCst) != 0) {}
+        var c: u32 = undefined;
+        // This need not be strong because of the loop that follows.
+        // TODO implement mutex3 from https://www.akkadia.org/drepper/futex.pdf in x86 assembly.
+        if (@cmpxchgWeak(u32, &self.lock, 0, 1, AtomicOrder.Acquire, AtomicOrder.Monotonic)) |value1| {
+            c = value1;
+            while (true) {
+                if (c == 2 or
+                    @cmpxchgWeak(u32, &self.lock, 1, 2, AtomicOrder.Acquire, AtomicOrder.Monotonic) == null)
+                {
+                    if (builtin.os == builtin.Os.linux) {
+                        _ = linux.futex_wait(@ptrToInt(&self.lock), linux.FUTEX_WAIT | linux.FUTEX_PRIVATE_FLAG, 2, null);
+                    } else {
+                        // spin-lock
+                    }
+                }
+                if (@cmpxchgWeak(u32, &self.lock, 0, 2, AtomicOrder.Acquire, AtomicOrder.Monotonic)) |value2| {
```

Comment:
Why does this jump straight to the "locked with waiters" state? shouldn't it just go back and try the original cmpxchg again?